test/e2e/chaos/basic.go

```diff
@@ -1640,151 +1637,224 @@ selector:
-				}
-				allSlowConnection := func() [][]int {
-					var result [][]int
-					for source := 0; source < len(networkPeers); source++ {
-						for target := source + 1; target < len(networkPeers); target++ {
-							delay := testDelay(source, target)
-							klog.Infof("delay from %d to %d: %d", source, target, delay)
-							if delay > 100*1e6 {
-								result = append(result, []int{source, target})
-							}
-						}
-					}
-
-					return result
-				}
+				var (
+					noExcepts = []*exceptMap{}
+				)
 
-				framework.ExpectEqual(len(allSlowConnection()), 0)
+				framework.ExpectEqual(len(allSlowConnection(c, networkPeers, ports, noExcepts)), 0)
 
-				// normal delay chaos
-				networkDelay := &v1alpha1.NetworkChaos{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      "network-chaos-1",
-						Namespace: ns,
-					},
-					Spec: v1alpha1.NetworkChaosSpec{
-						Action: v1alpha1.DelayAction,
-						Selector: v1alpha1.SelectorSpec{
-							Namespaces:     []string{ns},
-							LabelSelectors: map[string]string{"app": "network-peer-0"},
+				var (
+					testDelayTcParam = v1alpha1.TcParameter{
+						Delay: &v1alpha1.DelaySpec{
+							Latency:     "200ms",
+							Correlation: "25",
+							Jitter:      "0ms",
 						},
-						Mode: v1alpha1.OnePodMode,
-						TcParameter: v1alpha1.TcParameter{
-							Delay: &v1alpha1.DelaySpec{
-								Latency:     "200ms",
-								Correlation: "25",
-								Jitter:      "0ms",
-							},
+					}
+					testDelayTcParamEvenMoreComplicate = v1alpha1.TcParameter{
+						Delay: &v1alpha1.DelaySpec{
+							Latency:     "200ms",
+							Correlation: "25",
+							Jitter:      "0ms",
 						},
-						Duration: pointer.StringPtr("9m"),
-						Scheduler: &v1alpha1.SchedulerSpec{
-							Cron: "@every 10m",
+						Loss: &v1alpha1.LossSpec{
+							Loss:        "25",
+							Correlation: "25",
 						},
-					},
-				}
+						Duplicate: &v1alpha1.DuplicateSpec{
+							Duplicate:   "25",
+							Correlation: "25",
+						},
+						Corrupt: &v1alpha1.CorruptSpec{
+							Corrupt:     "25",
+							Correlation: "25",
+						},
+					}
+					testDelayDuration      = pointer.StringPtr("9m")
+					testDelaySchedulerSpec = &v1alpha1.SchedulerSpec{
+						Cron: "@every 10m",
+					}
+				)
+
+				// normal delay chaos
+				networkDelay := makeNetworkDelayChaos(
+					ns, "network-chaos-1",
+					map[string]string{"app": "network-peer-0"},
+					map[string]string{}, // no target specified
+					v1alpha1.OnePodMode,
+					v1alpha1.OnePodMode,
+					testDelayTcParam,
+					testDelayDuration,
+					testDelaySchedulerSpec,
+				)
+
 				klog.Infof("Injecting delay for 0")
 				err = cli.Create(ctx, networkDelay.DeepCopy())
 				framework.ExpectNoError(err, "create network chaos error")
 				time.Sleep(5 * time.Second)
-				framework.ExpectEqual(allSlowConnection(), [][]int{{0, 1}, {0, 2}, {0, 3}})
+				framework.ExpectEqual(allSlowConnection(c, networkPeers, ports, noExcepts), [][]int{{0, 1}, {0, 2}, {0, 3}})
```

Comment:
This case still return result `[][]int{{0, 1}}`, and after check the logs:

```

[2020-10-20T07:38:17.102Z] [5] Oct 20 07:38:16.833: INFO: network-peer-2-6d4d849666-gsvbl  chaos-mesh-worker2  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:21 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:22 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:22 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:21 +0000 UTC  }]

[2020-10-20T07:38:17.102Z] [5] Oct 20 07:38:16.833: INFO: network-peer-1-66b96cb95b-njm9z  chaos-mesh-worker3  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:16 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:18 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:18 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 07:37:16 +0000 UTC  }]

```

I found both of `network-peer-0`, `network-peer-2`, `network-peer-3` are deployed on chaos-mesh-worker2, and `network-peer-1` deployed on another node `chaos-mesh-worker3`. I'm not sure if it's the root cause...

Reply:
At the latest build, I see the logs below:

```
[2020-10-20T08:53:14.750Z] [5] I1020 08:53:14.629320      23 basic.go:1691] Injecting delay for 0
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.643029      23 basic.go:1967] testing delay from network-peer-0-68599f7689-dm4p5 to network-peer-1-54bb55d857-cxdwt
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.643083      23 basic.go:2068] sending request to localhost:37948 with body: {"targetIP":"10.244.3.5"}
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.645379      23 basic.go:1973] delay from 0 to 1: 513566
[2020-10-20T08:53:20.007Z] [5] I1020 08:53:19.645395      23 basic.go:1967] testing delay from network-peer-0-68599f7689-dm4p5 to network-peer-2-65c886b4b7-94756
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.645405      23 basic.go:2068] sending request to localhost:37948 with body: {"targetIP":"10.244.1.9"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.647108      23 basic.go:1973] delay from 0 to 2: 308449
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.647133      23 basic.go:1967] testing delay from network-peer-0-68599f7689-dm4p5 to network-peer-3-5d4b896675-kzmkg
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.647144      23 basic.go:2068] sending request to localhost:37948 with body: {"targetIP":"10.244.3.7"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.848924      23 basic.go:1973] delay from 0 to 3: 200461654
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.848947      23 basic.go:1967] testing delay from network-peer-1-54bb55d857-cxdwt to network-peer-2-65c886b4b7-94756
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.848957      23 basic.go:2068] sending request to localhost:33231 with body: {"targetIP":"10.244.1.9"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.850735      23 basic.go:1973] delay from 1 to 2: 355391
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.850754      23 basic.go:1967] testing delay from network-peer-1-54bb55d857-cxdwt to network-peer-3-5d4b896675-kzmkg
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.850764      23 basic.go:2068] sending request to localhost:33231 with body: {"targetIP":"10.244.3.7"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.852280      23 basic.go:1973] delay from 1 to 3: 267228
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.852297      23 basic.go:1967] testing delay from network-peer-2-65c886b4b7-94756 to network-peer-3-5d4b896675-kzmkg
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.852307      23 basic.go:2068] sending request to localhost:36099 with body: {"targetIP":"10.244.3.7"}
[2020-10-20T08:53:20.008Z] [5] I1020 08:53:19.853873      23 basic.go:1973] delay from 2 to 3: 299827
[2020-10-20T08:53:20.008Z] [5] Oct 20 08:53:19.853: FAIL: Expected
[2020-10-20T08:53:20.008Z] [5]     <[][]int | len:1, cap:1>: [[0, 3]]
[2020-10-20T08:53:20.008Z] [5] to equal
[2020-10-20T08:53:20.008Z] [5]     <[][]int | len:3, cap:3>: [[0, 1], [0, 2], [0, 3]]
```

```
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-0-68599f7689-glgrr  chaos-mesh-worker2  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:57 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:59 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:59 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:53:57 +0000 UTC  }]
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-1-54bb55d857-z7fm5  chaos-mesh-worker3  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:02 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:04 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:04 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:02 +0000 UTC  }]
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-2-65c886b4b7-rkwdn  chaos-mesh-worker3  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:07 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:09 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:09 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:07 +0000 UTC  }]
[2020-10-20T08:55:18.941Z] [2] Oct 20 08:55:18.237: INFO: network-peer-3-5d4b896675-p2vh2  chaos-mesh-worker2  Running         [{Initialized True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:13 +0000 UTC  } {Ready True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:15 +0000 UTC  } {ContainersReady True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:15 +0000 UTC  } {PodScheduled True 0001-01-01 00:00:00 +0000 UTC 2020-10-20 08:54:13 +0000 UTC  }]
```

Okay, at this time, it's peer-3...

Reply:
We judged whether the target is set by comparing `.Spec.Target` with `nil`. If `.Spec.Target` is not nil but it contains no selector, it will `List` the pods under the namespace.

And you select one pod from the pods, so the result seems to be random.

To test `netem` without filter, please remove the `.Spec.Target` field (or set it to nil).

Reply:
OK, many thanks!